keps/sig-api-machinery/3937-declarative-validation/README.md

```diff
@@ -0,0 +1,1077 @@
+| enum values            | `+enum`                                              | `enum`                                                                             |
+| uniqueness             | `listType=set` (sets and map keys)                   | `x-kubernetes-list-type`                                                           |
+| regex matches          | `+pattern`                                           | `pattern`                                                                          |
+| cross field validation | `validation=rule:"{CEL expression}"`                 | `x-kubernetes-validations`                                                         |
+| [transition rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules)   | `validation=rule:"{CEL expression using oldSelf}"`   | `x-kubernetes-validations`                                                         |
+| special case: metadata name format   | `+metadataNameFormat={format name}`                  | `x-kubernetes-validations` (see "format" section below for details) |
+
+### CEL Validation
+
+CEL will play a major role in declarative validation by offering a way to
+declare validation rules for use cases that are too complex to be declared using
+JSON Schema value validations. Using CEL to validate APIs has been successfully
+demonstrated by CRD Validation Rules, which are on track for GA in 1.28.
+
+In addition to the `rule`, `message` and `messageExpression` available to CRD
+validation rules, declarative validation rules will also have a `reason` field,
+e.g.:
+
+```go
+  
+  type FizzBuzzSpec struct {
+    //+validation=rule:"!['badname2', 'badname2'].exists(notAllowed, self.contains(notAllowed))",reason=Forbidden
+    WidgetName string `json...`
+  }
+```
+
+We will need to extend our CEL libraries to make it possible to migrate all the
+validation rules that exist in the Kubernetes API today.
+
+- `isFormat() <bool>` and `validateFormat() <list<string>>` will be added to allow formats to be checked in CEL
+  expression and for format violations to be reported using `messageExpression: "self.validateFormat('ipv6')"`
+- IP and CIDR libraries will be added that allow for a wide range of IP (v4 and v6) checks to be performed.
+
+TODO: Flesh out the exact library functions we will to add.
+
+### Formats
+
+A significant portion of all validation rules in the API check that a field
+value conforms to a particular "format". A prominent example is `metadata.name`
+and `metadata.generateName` validation.
+
+We will extend the [available list of
+formats](https://github.com/kubernetes/kube-openapi/blob/7fbd8d59e5b89f2ca43a5dcececbffc0bb186c37/pkg/validation/strfmt/default.go#L128)
+to cover formats heavily used by the Kubernetes API, namely:
+
+| Format                       | Primary validate uses              |
```

Comment:
I'm glad this exists, we desperately need it. It'd be nice if we could publish/make official:
1. How these are parsed, so that it's kind of part of k8s api definition
2. For some of these, if we have a specific way to normalize them (same, a formula that can be used client side would be neat).

How can we document these properly? Maybe a reference in the kep is good enough, but then you should elaborate here possibly?